### PR TITLE
Feature/network incentivisation

### PR DESCRIPTION
**Pull request by cammellos**
Add interactions with the contract & posting pings on whisper.

If started with an incentivisation config it will:
- try to register with the contract
- post whisper envelopes on a specific channel, that can be read by status client advertising its address
- check with the contract which nodes are available and collects pings from them
- vote based on the results of the pings

Still in progress and much to do, but I would like to merge as we go.

### CODE REVIEW

File: services/incentivisation/contract.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+	tx, err := c.Vote(opts, joinNodes, removeNodes)
+	if err != nil {
+		return nil, err
+	}
+
+	for {
+		receipt, _ := c.client.TransactionReceipt(context.TODO(), tx.Hash())
+		if receipt != nil {
+			if receipt.Status == 0 {
+				return nil, errors.New("Receipt invalid")
```
Comment (pedropombeiro) [inline services/incentivisation/contract.go]: Shouldn't it be
```suggestion
				return nil, errors.New("Invalid receipt")
```

File: services/incentivisation/service.go [hunk 2]
```diff
@@ -0,0 +1,518 @@
...
+	privateKey      *ecdsa.PrivateKey
+	log             log.Logger
+	// The first round we will not be voting, as we might have incomplete data
+	initialSession uint64
+	// The current session
+	currentSession uint64
+	whisperPings   map[string][]uint32
+}
+
+// New returns a new Service
```
Comment (pedropombeiro) [inline services/incentivisation/service.go]:
```suggestion
// New returns a new incentivization Service
```

File: services/incentivisation/service.go [hunk 3]
```diff
@@ -0,0 +1,518 @@
...
+}
+
+// Protocols returns a new protocols list. In this case, there are none.
+func (s *Service) Protocols() []p2p.Protocol {
+	return []p2p.Protocol{}
+}
+
+// APIs returns a list of new APIs.
+func (s *Service) APIs() []rpc.API {
+	s.log.Info("APIS CALLED")
```
Comment (pedropombeiro) [inline services/incentivisation/service.go]: Why all-caps here? Is it a remnant?

File: services/incentivisation/service.go [hunk 4]
```diff
@@ -0,0 +1,518 @@
...
+	_, err = s.register()
+	if err != nil {
+		s.log.Error("error querying contract", "registered", err)
+		return err
+	}
+	return nil
+}
+
+// newSession checks if we are in a new session and updates the session if so
+func (s *Service) newSession() (bool, error) {
```
Comment (pedropombeiro) [inline services/incentivisation/service.go]: maybe `ensureSession` or `getSession` would better describe what the method does?

File: services/incentivisation/service.go [hunk 5]
```diff
@@ -0,0 +1,518 @@
...
+	}
+
+	err = s.fetchMessages()
+	if err != nil {
+		return err
+	}
+
+	err = s.checkRegistered()
+	if err != nil {
+		s.log.Error("Could not check registered", "err", err)
```
Comment (pedropombeiro) [inline services/incentivisation/service.go]:
```suggestion
		s.log.Error("Could not check if node is registered with the contract", "err", err)
```

File: services/incentivisation/service.go [hunk 6]
```diff
@@ -0,0 +1,518 @@
...
+			case <-s.quit:
+				s.ticker.Stop()
+				return
+			}
+		}
+	}()
+}
+
+func (s *Service) Start(server *p2p.Server) error {
+	s.log.Info("Incentivisation service started 2", "address", s.addressString(), "publickey", s.publicKeyString())
```
Comment (pedropombeiro) [inline services/incentivisation/service.go]: What is the 2 in `Incentivisation service started 2`?

File: services/incentivisation/service.go [hunk 7]
```diff
@@ -0,0 +1,518 @@
...
+
+	response, err := s.contract.RegisterNode(&bind.TransactOpts{
+		GasLimit: gasLimit,
+		From:     auth.From,
+		Signer:   auth.Signer,
+	}, s.publicKeyBytes(), ip, s.config.Port)
+	if err != nil {
+		return false, err
+	}
+	s.log.Info("resposne from registered", "response", response)
```
Comment (pedropombeiro) [inline services/incentivisation/service.go]:
```suggestion
	s.log.Info("response from registered", "response", response)
```

File: services/incentivisation/service.go [hunk 8]
```diff
@@ -0,0 +1,518 @@
...
+}
+
+func (s *Service) FetchEnodes() error {
+	one := big.NewInt(1)
+
+	activeNodeCount, err := s.contract.ActiveNodeCount(nil)
+	if err != nil {
+		return err
+	}
+	s.log.Info("fetched node count", "count", activeNodeCount)
```
Comment (pedropombeiro) [inline services/incentivisation/service.go]:
```suggestion
	s.log.Info("fetched active node count", "count", activeNodeCount)
```

File: services/incentivisation/service.go [hunk 9]
```diff
@@ -0,0 +1,518 @@
...
+		if node.PublicKeyString() != s.publicKeyString() {
+			s.nodes[node.PublicKeyString()] = node
+		}
+	}
+
+	inactiveNodeCount, err := s.contract.InactiveNodeCount(nil)
+	if err != nil {
+		return err
+	}
+	s.log.Info("fetched node count", "count", inactiveNodeCount)
```
Comment (pedropombeiro) [inline services/incentivisation/service.go]:
```suggestion
	s.log.Info("fetched inactive node count", "count", inactiveNodeCount)
```

**Review by pedropombeiro — APPROVED** [review]
Lgtm, added some minor suggestions

File: services/incentivisation/service.go (hunk 3 above)
Comment (cammellos) [reply services/incentivisation/service.go]: yes, will remove, thanks!

File: services/incentivisation/service.go (hunk 6 above)
Comment (cammellos) [reply services/incentivisation/service.go]: testing docker images :) I will remove it

**Pull request merged**